libbeat/kibana/transform.go

```diff
@@ -120,63 +125,74 @@ func transformField(f common.Field) (common.MapStr, common.MapStr) {
 
 		if f.Format != "" {
 			format["id"] = f.Format
-			addStringFormatParam(&format, "inputFormat", f.InputFormat)
-			addStringFormatParam(&format, "outputFormat", f.OutputFormat)
-			addIntPFormatParam(&format, "outputPrecision", f.OutputPrecision)
-			addStringFormatParam(&format, "labelTemplate", f.LabelTemplate)
-			addStringFormatParam(&format, "urlTemplate", f.UrlTemplate)
 		}
-		addStringFormatParam(&format, "pattern", f.Pattern)
+		addParams(&format, version, f)
 	}
 
 	return field, format
 }
 
-func setVal(f common.MapStr, attr string, p *bool, def bool) {
-	if p != nil {
-		f[attr] = *p
-	}
-	if f[attr] == nil {
-		f[attr] = def
+func getVal(valP *bool, def bool) bool {
+	if valP != nil {
+		return *valP
 	}
+	return def
 }
 
-func setAggregatable(f common.MapStr, c common.Field) {
-	attr := "aggregatable"
-	if c.Aggregatable != nil {
-		f[attr] = *c.Aggregatable
-	} else if c.Type == "text" {
-		f[attr] = false
-	} else if f[attr] == nil {
-		f[attr] = true
-	}
+func addParams(format *common.MapStr, version *common.Version, f common.Field) {
+	addFormatParam(format, "pattern", f.Pattern)
+	addFormatParam(format, "inputFormat", f.InputFormat)
+	addFormatParam(format, "outputFormat", f.OutputFormat)
+	addFormatParam(format, "outputPrecision", f.OutputPrecision)
+	addFormatParam(format, "labelTemplate", f.LabelTemplate)
+	addVersionedFormatParam(format, version, "urlTemplate", f.UrlTemplate)
 }
 
-func getVal(valP *bool, def bool) bool {
-	if valP != nil {
-		return *valP
+func addFormatParam(f *common.MapStr, key string, val interface{}) {
+	switch val.(type) {
+	case string:
+		if v := val.(string); v != "" {
+			createParam(f)
+			(*f)["params"].(common.MapStr)[key] = v
+		}
+	case *int:
+		if v := val.(*int); v != nil {
+			createParam(f)
+			(*f)["params"].(common.MapStr)[key] = *v
+		}
 	}
-	return def
 }
 
-func addStringFormatParam(f *common.MapStr, key string, val string) {
-	if val == "" {
+// takes the highest version where major version <= given version
+func addVersionedFormatParam(f *common.MapStr, version *common.Version, key string, val []common.VersionizedString) {
+	if len(val) == 0 {
 		return
 	}
-	if (*f)["params"] == nil {
-		(*f)["params"] = common.MapStr{}
+	paramVer, _ := common.NewVersion("0.0.0")
+	var paramVal string
+	for _, v := range val {
+		minVer, err := common.NewVersion(v.MinVersion)
```

Comment:
This case would mean an invalid fields file? Should we at least log an error?

Reply:
I have given it a second thought and I think it is even better to stop the processing here and return an error, since the kibana pattern can't be created as intended.